### DISCUSSION TITLE
Contextual platform

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#12250] [Ideas]
## Problem

A Dagger function can't access its client's platform (OS and arch). As a result, the only way to build a native binary for the client, is to request mandatory arguments.

For example:

```console
dagger call native-build --os=$(uname -s) --arch=$(uname -m) -o .
```

Instead I would like to do:

```console
dagger call native-build -o .
```

## Solution

Apply the same "context access" model as for contextual directories:

- Function can receive a `dagger.Platform` as argument
- If the argument is optional, the default value is the client's platform

Example:

```golang
func (m *MyModule) Build(
ctx context.Context,
source *dagger.Directory
// +optional
platform *dagger.Platform,
) {
  // We assume there is a reusable go module that handles platform
  return dag.Go.Build(source, platform)
}
```

**Comment by helderco** [comment]
> [...] the only way to build a native binary for the client, is to request mandatory arguments.

The CLI supports a simpler way since https://github.com/dagger/dagger/pull/7158 by using the value "current":

```console
dagger call native-build --platform=current -o .
```

**Comment by shykes** [comment]
> > [...] the only way to build a native binary for the client, is to request mandatory arguments.
> 
> The CLI supports a simpler way since #7158 by using the value "current":
> 
> ```
> dagger call native-build --platform=current -o .
> ```

Wow I had no idea! I guess that will make this feature even easier to implement, if part of the work is already done? 🤞

**Comment by marcosnils** [comment]
resurfacing this issue with more context from @shykes in discord here: 

https://discord.com/channels/707636530424053791/1285660387433513114/1285671835454673040

> I know 🙂 I want a way to choose between them in my pragmas
Proposal:
```go
func Foo(
  // +optional
  // +default="client"      <-- OPTION 1: use the client's platform
  // +default="runtime"    <-- OPTION 2: use the current runtime platform (running the current function)
  // +default="engine"      <-- OPTION 3: use the engine's default platform
  platform dagger.Platform,
)
```

> The CLI would map them:

> $ dagger call foo --platform=client
> $ dagger call foo --platform=runtime
> $ dagger call foo --platform=engine


> "runtime" and "engine" may not be the same if eg. the engine supports multiple platforms, and the current function is running in a non-default one, in qemu for some reason. I don't think there's a way to configure this at the moment, but it could conceivably happen

**Comment by jpadams** [comment]
Watching with interest. Ran into need for this today.

**Comment by shykes** [comment]
I think I might just implement this... Tired of waiting

**Comment by nipuna-perera** [comment]
Needed this today. Intuitively, I tried to use `dag.Module().Runtime().Platform(ctx)` but that panic'ed. 

A use case for me would be building a docker image for the architecture of the host platform. Need a way to detect host platform or engine runtime platform which would in most cases match host platform.

**Comment by jedevc** [comment]
There's one important thing I want to resolve - this propagates *all the way back to the client*? I just wonder if that's the behavior we want, or whether we just want the platform of the caller.

So if I go `dagger shell` calls module `A` calls module `B` calls module `C` - module `C` is able to access the client platform, i.e. `dagger shell`? This is a bit of a departure from how we call and cache things today: today the only thing that matters is `B` calling `C`. If we allow this, it would be the first "global variable" we'd have. Note: context directories are different than this, because they read from the *called* module context, not the *caller*.

Personally, I think we should allow these special values for now (don't really care about the names):
- `current`/`client`
  - the platform that the module is actually running in
- `caller`/`engine`/`runtime`
  - the platform that the caller is running: if called from the cli/`shell`, it's the platform of the cli, if it's called from another module, then it's actually the same as `current` (but because it's being called from code, then the code should manually propagate the value here).

That said, maybe there is value in a `client` value that's from the top-most level? We just need to be careful that caching keeps working properly (and that we don't cache calls to the hypothetical module `C` when the client platform is different).

Could we defer on adding a `client` value that reaches back to the top-level for now? The others should be super simple to implement, we could ship them in the next release, I just have a couple reservations on that value specifically that I think we should bikeshed while getting something concrete shipped.

---

Implementation notes for anyone picking this up:
- We should be able to handle this entirely in https://github.com/dagger/dagger/blob/20bab132591af1b9401b9d84a0d5a8b231525ec9/core/platform.go#L56-L67
- To get the `current` platform, we just need to return the `platforms.Default()` - all modules run on the same architecture as the engine today, we don't need to make this more complicated right now.
- To get the `caller` platform is trickier.
	- We should have this information attached through the `ClientMetadata` type, so when we create those, we can attach the `platforms.Default()`.
	- But `DecodeInput` doesn't currently take a `context.Context` - so we'll need to change the function signature to allow that.

**Comment by jedevc** [comment]
Bleh, I started hacking in https://github.com/dagger/dagger/compare/main...jedevc:dagger:special-platform-values to see if the approach works, turns out to be a bit trickier than I maybe thought.

The problem is that this is a "dynamic" default type - dynamic based on whose calling, etc. I think this is possible to handle? But I think it still requires a bit of hackery to make sure we don't screw up the cache (we'd need to switch to resolving default args before computing cache keys, for example).

It also means we need to potentially delay decoding the default input value - which is fiddly. We'll also need to do some messing around to make sure that the docs are actually right (and preserve the special values, instead of defaulting to the client that generates the docs).

(note, we'd need to solve all the above problems anyways to do more contextual stuff of any sort really, like `// +env`, etC)
